### DISCUSSION TITLE
Webview timer (setTimeout, setInterval) throttling

### DISCUSSION THREAD

**Discussion by matthewjamesadam** [microsoft/vscode-discussions#983] [Extension Development QnA] [answered]
Hello friends -- does anyone know if webviews have any kind of timer throttling that is sometimes enabled?

We have a webview that on launch can do quite a bit of compute-intensive work -- the content it is rendering is streamed from a remote source, and we render the content as it is comes in.  A simplified version of our the main loop looks something like:

```
async UpdateFunction() {
  let done = false;
  while (!done) {
    const nextValue = getNextValueFromStream();
    if (nextValue) {
      updateValueInDOM(nextValue);
      await new Promise((resolve) => setTimeout(resolve, 50)); // Wait for 50 ms, before updating the next item
    }
    else {
      done = true; // No more items, all done
    }
  }
}
```

The idea is that the view progressively updates values until the stream is complete.  This doesn't block due to the awaiting sleep.

This generally works, but every now and then I am seeing very odd behaviour: `setTimeout(resolve, 50)` does not wait for 50ms, it waits for a full second, on every iteration.  The odd thing is that this seems to occur on a per-webview basis: every 20 or 30 times I run this webview, every single `setTimeout` in that particular webview only triggers once a second.  Other `setTimeout` and `setInterval` invocations in that webview seem to be similarly blocked.  This continues until the webview is shut down, and then the next webview to be launched appears to work correctly.

I captured the javascript VM behaviour in the performance tab, and didn't notice anything unusual.  There are spikes of work being done every second, as expected, and no work being done in between.

I see that Chrome has some timer throttling (https://developer.chrome.com/blog/timer-throttling-in-chrome-88), and electron has some kind of throttling as well, but I'm not sure if either of those is relevant to a VSCode extension webview or not.

Thanks!

**Comment by r3m0t** [comment, marked as answer]
Yes, VSCode Electron is almost identical to Electron and Chrome.

Take a look at `requestAnimationFrame`, `requestIdleCallback`, or moving work to a Service Worker to keep the page responsive. Also, consider calling `getNextValueFromStream` multiple times before doing `await`. You can use `performance.hrtime` to check whether your calculation is taking too long, at which point you should stop running your JS to let the browser run other things.

**Reply by matthewjamesadam** [reply]
Thanks @r3m0t -- I did try the animation frame timer (same behaviour).  I will see if the idle callback timer has any differing behaviour.  I considered a service worker, but it seemed unnecessary since the actual compute work being done locally is so minimal.  This is within a webview, so node's `performance.hrtime` is not relevant.

Right now we indeed are doing the equivalent of calling `getNextValueFromStream` repeatedly, which ensures we consume all of the pending data -- the problem is that whenever we relinquish execution (ie, waiting for a timer or other async operation), we never get called back for up to a full second.  This results in chunky rendering, as ideally we'd be updating with any pending data much quicker (every 50ms, for instance).

**Reply by r3m0t** [reply]
Right, I meant https://developer.mozilla.org/en-US/docs/Web/API/DOMHighResTimeStamp
